Add PerceiverForTokenClassification

NielsRogge opened the pull request:
# What does this PR do?

You can also use the Perceiver to do named-entity recognition! Training for 3 epochs on CoNLL 2003 gives me the following result (using the official example script):

```
***** eval metrics *****
  epoch                   =        3.0
  eval_LOC_f1             =     0.8895
  eval_LOC_number         =      11884
  eval_LOC_precision      =     0.8934
  eval_LOC_recall         =     0.8856
  eval_MISC_f1            =     0.8058
  eval_MISC_number        =       6078
  eval_MISC_precision     =     0.8182
  eval_MISC_recall        =     0.7938
  eval_ORG_f1             =     0.8215
  eval_ORG_number         =       8869
  eval_ORG_precision      =     0.8187
  eval_ORG_recall         =     0.8243
  eval_PER_f1             =     0.8477
  eval_PER_number         =      10479
  eval_PER_precision      =     0.8598
  eval_PER_recall         =     0.8359
  eval_loss               =     0.1479
  eval_overall_accuracy   =     0.9594
  eval_overall_f1         =      0.848
  eval_overall_precision  =     0.8539
  eval_overall_recall     =     0.8421
  eval_runtime            = 0:01:25.54
  eval_samples            =       3250
  eval_samples_per_second =      37.99
  eval_steps_per_second   =      4.757
```

This PR includes the tweaked example script to make it work for the Perceiver. I can remove it if required, or move it to research-projects directory.

Fixes #14971

NielsRogge committed:
First draft

NielsRogge committed:
Add script

NielsRogge committed:
Fix script

NielsRogge committed:
Add arguments

NielsRogge committed:
Add print statement for debugging

NielsRogge committed:
Fix data collator

NielsRogge committed:
Remove print statements

NielsRogge committed:
Fix code quality

sgugger commented on examples/pytorch/token-classification/run_ner_perceiver.py:
```diff
@@ -0,0 +1,462 @@
+            truncation=True,
+            # We use this argument because the texts in our dataset are lists of words (with a label for each word).
+            is_split_into_words=True,
+            # return_offsets_mapping=True,
```
This line should be removed.

sgugger commented on examples/pytorch/token-classification/run_ner_perceiver.py:
```diff
@@ -0,0 +1,462 @@
+            # -2 special token [BOS]  [SEP]
+            id_length = len(tokenized_inputs["input_ids"][i]) - 2
+            char_length = len("".join(examples[text_column_name][i]))
+            assert id_length == char_length
```
Not sure the empty asserts will help the user here. At least an error message to guide them in case of error?

sgugger commented on src/transformers/models/perceiver/modeling_perceiver.py:
```diff
@@ -174,6 +174,39 @@ class PerceiverClassifierOutput(ModelOutput):
+@dataclass
+class PerceiverTokenClassifierOutput(ModelOutput):
+    """
+    Args:
+    Base class for Perceiver's outputs of token classification models.
+        loss (`torch.FloatTensor` of shape `(1,)`, *optional*, returned when `labels` is provided) :
+            Classification loss.
+        logits (`torch.FloatTensor` of shape `(batch_size, sequence_length, config.num_labels)`):
+            Classification scores (before SoftMax).
+        hidden_states (`tuple(torch.FloatTensor)`, *optional*, returned when `output_hidden_states=True` is passed or
+        when `config.output_hidden_states=True`):
+            Tuple of `torch.FloatTensor` (one for the output of the embeddings + one for the output of each layer) of
+            shape `(batch_size, sequence_length, hidden_size)`. Hidden-states of the model at the output of each layer
+            plus the initial embedding outputs.
+        attentions (`tuple(torch.FloatTensor)`, *optional*, returned when `output_attentions=True` is passed or when
+        `config.output_attentions=True`):
+            Tuple of `torch.FloatTensor` (one for each layer) of shape `(batch_size, num_heads, sequence_length,
+            sequence_length)`. Attentions weights after the attention softmax, used to compute the weighted average in
+            the self-attention heads.
+        cross_attentions (`tuple(torch.FloatTensor)`, *optional*, returned when `output_attentions=True` is passed or
+        when `config.output_attentions=True`):
+            Tuple of `torch.FloatTensor` (one for each layer) of shape `(batch_size, num_heads, sequence_length,
+            sequence_length)`. Attentions weights of the decoder's cross-attention layer, after the attention softmax,
+            used to compute the weighted average in the cross-attention heads.
```
This docstring needs to be rewritten. It has been badly formatted by the doc-styler since the line "Base class for Perceiver's outputs of token classification models." is after "Args:" instead of before.

sgugger reviewed (approved):
Thanks a lot for adding this! Not too sure about the example. It can go with the maintained ones (with proper documentation in the README) since it's a small change (and could work for any character-based tokenizer from what I see).
Wdyt @LysandreJik ?

NielsRogge commented:
cc'ing @Narsil as the token classification pipeline currently doesn't work with the Perceiver. Shoud I remove it from the Auto API?

NielsRogge committed:
Apply suggestions from code review

Narsil commented:
@NielsRogge What does happen when it doesn't work ?

Is it taking text only ? then it should be fixable, no ?

Narsil committed:
Making the pipeline work.

- `scores.shape` > `input_ids.shape` most of the time
- We simply need to relax the assumption `scores.shape ==
input_ids.shape` that the code was making.
